File: 547. Friend Circles.py
class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        # DFS approach
        N = len(M)
        visited = [False] * N
        count = 0
        for i in range(N):
            if visited[i]:
                continue
            count += 1
            self.dfs(M, i, visited)            
        return count
            
    def dfs(self, M, i, visited):
        visited[i] = True
        for j in range(len(M[i])):
            if M[i][j] == 1 and not visited[j]:
                self.dfs(M, j, visited)
     

File: 547. Friend Circles.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& M) {
        int count = 0;
        vector<int> visited(M.size(), 0);
        for (int i = 0; i != M.size(); ++i) {
            if (!visited[i]) {
                count++;
                dfs(M, i, visited);
            }
        }
        return count;
    }
    
    void dfs(vector<vector<int>>& M, int i, vector<int>& visited) {
        visited[i] = 1;
        for (int j = 0; j != M[i].size(); ++j) {
            if (M[i][j] && !visited[j])
                dfs(M, j, visited);
        }
    }
};
